iptables/src/utils.cpp

```diff
@@ -0,0 +1,309 @@
+/**
+ * Copyright (c) 2018 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *  Parts of this file are also:
+ *
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <osquery/sdk.h>
+
+#include <net/if.h>
+#include <netdb.h>
+#include <string>
+#include <sys/socket.h>
+
+#include <boost/algorithm/string/trim.hpp>
+
+#include <trailofbits/extutils.h>
+
+#include "utils.h"
+
+using namespace osquery;
+
+namespace trailofbits {
+static const std::string kIptablesSave = "/sbin/iptables-save";
+static const std::string kIp6tablesSave = "/sbin/ip6tables-save";
+
+static const std::string kIptablesNamesFile = "/proc/net/ip_tables_names";
+static const std::string kIp6tablesNamesFile = "/proc/net/ip6_tables_names";
+
+static const std::string kHexMap = "0123456789ABCDEF";
+
+static const int kMaskHighBits = 4;
+static const int kMaskLowBits = 15;
+
+static TableList getTableNames(const std::string& filename) {
+  std::string content;
+
+  auto s = readFile(filename, content);
+
+  if (!s.ok()) {
+    TLOG << "Error reading: " << filename << ": " << s.toString();
+    return {};
+  }
+
+  auto results = SplitString(content, '\n');
+
+  for (auto& result : results) {
+    boost::algorithm::trim(result);
+  }
+
+  return results;
+}
+
+/* NOTE(ww): This function takes either iptables-save or ip6tables-save.
+ * It would be nice to use iptables-xml and ip6tables-xml, but ip6tables-xml
+ * doesn't exist yet and iptables-xml frequently hangs (during experimentation).
+ */
+Status genMatchMap(const std::string& ipt_cmd, MatchMap& match_map) {
+  ProcessOutput output;
+
+  if (!ExecuteProcess(output, ipt_cmd, {}) || output.exit_code != 0) {
+    return Status(1, "couldn't exec " + ipt_cmd);
+  }
+
+  if (output.std_output.empty()) {
+    return Status(1, "no output from command");
+  }
+
+  std::string table;
+  std::string chain;
+  for (auto line : SplitString(output.std_output, '\n')) {
+    std::string match;
+    std::string target;
+    std::string target_options;
+
+    // If the line is a comment, skip it.
+    if (line.at(0) == '#') {
```

Comment:
this will raise an exception if line is an empty string -- perhaps we want to detect empty strings and simply skip them? although i guess it should be technically impossible for line to be an empty string, it should always at least have a `\n`

Reply:
Yeah, I'll add this check.



> although i guess it should be technically impossible for line to be an empty string, it should always at least have a `\n`

Unfortunately(?) not, since `SplitString` trims each line:

```cpp
std::vector<std::string> SplitString(const std::string& buffer,
                                     char delimiter) {
  std::istringstream stream(buffer);

  std::vector<std::string> output;
  for (std::string str; std::getline(stream, str, delimiter);) {
    boost::algorithm::trim(str);
    output.push_back(str);
  }

  return output;
}
```

Reply:
the boost docs for trim mention that 'spaces' are removed, but i'm not sure about whitespace in general? that what made me think that

i really should set up a minimal boost environment to test things like this..

https://www.boost.org/doc/libs/1_66_0/doc/html/boost/algorithm/trim.html

Reply:
Ah, hmm. A quick SO search says that boost's `trim` removes newlines as well, but I've added a check in any case.